### ISSUE TITLE
option for gradient accumulation in wav2letter?

### ISSUE THREAD

**Issue by trangtv57** [facebookresearch/wav2letter#649]
I can not find gradient accumulation option in training flags. Does wav2letter support this feature?

**Comment by padentomasello** [comment]
Hi @trangtv57, we do not currently have a training flag to do gradient accumulation. This shouldn't be too difficult to implement however if you want to give it a shot. 

If you take a look at https://github.com/facebookresearch/wav2letter/blob/master/Train.cpp#L630-L668, this is where we zero the gradients with (`opt->zeroGrad()`), compute the new gradients with `loss.backward()`, and update the weights with `opt->step()`.  (where opt is both the net and criterion optimizer)

For gradient accumulation, we can move `opt->zeroGrad()` and `opt->step()` to happen only after a certain number of batches. (`loss.backward()` will automatically sum up the gradients from each of the steps)

You'll also need to make sure you're scaling to gradients correctly: https://github.com/facebookresearch/wav2letter/blob/master/Train.cpp#L642.  (You'd probably want to only do this before updating the weights, and scale the gradients by the "accumulated" batch size). 

Let me know if this makes sense or if you have any questions.

**Comment by trangtv57** [comment]
thanks padento,
I understand your instruction, so I change my code from line 630->668 like this:
`loss.backward();
        // 2 mean number accumulation step
        if (curBatch % 2 == 0){
            netopt->zeroGrad();
            critopt->zeroGrad();

            if (reducer) {
              reducer->finalize();
            }
            af::sync();
            meters.bwdtimer.stopAndIncUnit();

            // optimizer
            meters.optimtimer.resume();

            // scale down gradients by batchsize
            for (const auto& p : ntwrk->params()) {
              if (!p.isGradAvailable()) {
                continue;
              }
              // *2 mean accumulation step
              p.grad() = p.grad() / (FLAGS_batchsize * 2);
            }
            for (const auto& p : crit->params()) {
              if (!p.isGradAvailable()) {
                continue;
              }
              // *2 mean accumulation step
              p.grad() = p.grad() / (FLAGS_batchsize * 2);
            }

            // clamp gradients
            if (FLAGS_maxgradnorm > 0) {
              auto params = ntwrk->params();
              if (clampCrit) {
                auto critparams = crit->params();
                params.insert(params.end(), critparams.begin(), critparams.end());
              }
              fl::clipGradNorm(params, FLAGS_maxgradnorm);
            }

            // update weights
            critopt->step();
            netopt->step();
        }`
with 2 is fix accumulation step (because i don't know how to set this as option flags). 
I can build and run it normal, but can you confirm does it work correctly?
One more thing, I use curBatch as count for step but I think curBatch variable mean global step and I need add 1 variable for local step in loop, and curBatch increase (++) when i try opt->step like this: 
line from 547->550 (start load each batch for train)
`      int localBatchStep = 0;
      for (auto& batch : *trainset) {
        localBatchStep++;`
and line from 630 -> 670 change with ++curBatch  inside when accumulation like this:
`      loss.backward();
        // 2 mean number accumulation step
        if (localBatchStep % 2 == 0){
            netopt->zeroGrad();
            critopt->zeroGrad();

            if (reducer) {
              reducer->finalize();
            }
            af::sync();
            meters.bwdtimer.stopAndIncUnit();

            // optimizer
            meters.optimtimer.resume();

            // scale down gradients by batchsize
            for (const auto& p : ntwrk->params()) {
              if (!p.isGradAvailable()) {
                continue;
              }
              // *2 mean accumulation step
              p.grad() = p.grad() / (FLAGS_batchsize * 2);
            }
            for (const auto& p : crit->params()) {
              if (!p.isGradAvailable()) {
                continue;
              }
              // *2 mean accumulation step
              p.grad() = p.grad() / (FLAGS_batchsize * 2);
            }

            // clamp gradients
            if (FLAGS_maxgradnorm > 0) {
              auto params = ntwrk->params();
              if (clampCrit) {
                auto critparams = crit->params();
                params.insert(params.end(), critparams.begin(), critparams.end());
              }
              fl::clipGradNorm(params, FLAGS_maxgradnorm);
            }

            // update weights
            ++curBatch;
            critopt->step();
            netopt->step();
        }`
thank you.

**Comment by trangtv57** [comment]
sorry because all code not in component?
I have another question about lexicon for training phase. With wav2letter component, all token represent by all phoneme like: 
down	d o u n  |
into	ˈi n t o͞o |
so some oov word such as intowdfe: can read in many way like: 
intowdfe   i n t o͞o u d
intowdfe  i n t o͞o u
so can wav2letter can handle multi representation like my idea, or can't do it and just can use lexicon in phase decoding with lm.

**Comment by trangtv57** [comment]
hey @padentomasello, can you answer me for close this issues?
I just want confirm for accumulation code.
Thanks you again.

**Comment by padentomasello** [comment]
hi @trangtv57 the code looks good to me! If you want to sanity check it though, you could try inspecting the gradients of a parameter after using gradient accumulation with a `accumulate_steps = 2` with double the batch size. The gradients should be the same. (You can do this after just one update). 

Please open a separate issue for the lexicon question. Thanks!
